### PR TITLE
Add example for iterator_flatten

### LINKED ISSUE #82687: Documentation of Iterator `flatten()` improvement
While linting my code I came across [`manual_flatten`](https://rust-lang.github.io/rust-clippy/master/index.html#manual_flatten) which was triggered in my code.

This suggests the change from:
```rust
for n in x {
    if let Some(n) = n {
        // something here
    }
}
```
Into:
```rust
for n in x.into_iter().flatten() {
    // something here
}
```
Both are valid code with same result. 

But the [rust docs on `flatten()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.flatten) does not state that this is expected behavior. Yes, this is stated [here](https://doc.rust-lang.org/std/option/enum.Option.html#method.flatten).
But this was not strait forward (in my opinion) to find. It took me writing a whole bug report and looking a bunch of thing up to figure this out. (and I'm not even new to Rust anymore)

I would suggest adding some example code to make his explicit. Here is a suggestion:
(after the ["Mapping and then flattening: `<code>...</code>`"](https://doc.rust-lang.org/std/iter/trait.Iterator.html#examples-22) section)
> Flattening also works on other types like `Option` and `Result`:
> ```rust
> let students_in_seats = vec![Some("Hasan"), Some("Sarah"), None, Some("Tim")];
> let students_present: Vec<_> = students_in_seats.into_iter().flatten().collect();
> assert_eq!(students_present, vec!["Hasan", "Sarah", "Tim"]);
> ```
(or something similar)

This makes both this behavior cleared, easier to find and exposes people to code like this so they are less surprised if clippy starts warning them.

It looks like this behavior might be more common in the future too.
https://github.com/rust-lang/rust-clippy/issues/6061
https://github.com/rust-lang/rust-clippy/issues/6676

### PR DESCRIPTION
**Pull request by HintringerFabian**
Adds an Example to iterator_flatten
Fixes #82687

### CODE REVIEW

File: library/core/src/iter/traits/iterator.rs [hunk 1]
```diff
@@ -1495,6 +1495,14 @@ pub trait Iterator {
     /// assert_eq!(merged, "alphabetagamma");
     /// ```
     ///
+    /// Flattening also works on other types like Option and Result:
```
Comment (the8472) [inline library/core/src/iter/traits/iterator.rs:1498]: It's not really specific to `Option` or `Result` though. It's that `Flatten` doesn't require `Item: Iterator`, it only needs  `Item: IntoIterator` and Option/Result are types that implement that.

Comment (HintringerFabian) [reply library/core/src/iter/traits/iterator.rs:1498]: I think i do not really understand what you mean, should we not add this example should i change it?

Comment (the8472) [reply library/core/src/iter/traits/iterator.rs:1498]: I'm pointing out that "types like Option and Result" is very vague. It doesn't tell the reader which those types have in common so they can be used with Flatten. I don't have a particular improvement in mind, I was just pointing out the vagueness and how it actually works.

Comment (Mark-Simulacrum) [reply library/core/src/iter/traits/iterator.rs:1498]: It seems like we already document that "iterator of iterators or an iterator of things that can be turned into iterators" above -- I'm not sure another example is all that useful. I think we have enough examples personally for flatten.

Comment (cuviper) [reply library/core/src/iter/traits/iterator.rs:1498]: I think this is okay, although its placement is interrupting the "you can also rewrite this" example below, so we should move this down. For wording, perhaps:

```suggestion
    /// Flattening works on any `IntoIterator` type, including `Option` and `Result`:
```

Although we're not *showing* `Result`, so maybe do so too, or remove that part.

**Review by cuviper — CHANGES_REQUESTED** [review]
Your new example is still interrupting the two "alphabetagamma" examples -- please move yours down, and then I think this looks good!

**Pull request merged**